File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
private:
    priority_queue<int> max_heap;
    priority_queue<int, vector<int>, greater<int>> min_heap;
public:
    MedianFinder() {
    }
    
    void addNum(int num) {
        // TC- O(logn)
        // SC- O(n)
        if(max_heap.empty() or num<max_heap.top())
            max_heap.push(num);
        else    
            min_heap.push(num);
        
        // balance the heaps
        if(min_heap.size()>max_heap.size()){
            max_heap.push(min_heap.top());
            min_heap.pop();
        }
        else if(min_heap.size()+1<max_heap.size()){
            min_heap.push(max_heap.top());
            max_heap.pop();
        }
    }
    
    double findMedian() {
        // TC- O(1)
        // SC- O(1)
        if(min_heap.size()==max_heap.size())
            return (double)(min_heap.top()+max_heap.top())/2.0;
        else
            return max_heap.top();
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.min_heap = []
        self.max_heap = []

    def addNum(self, num: int) -> None:
        # TC- O(logn)
        if not self.max_heap or num<-self.max_heap[0]:
            heapq.heappush(self.max_heap, -num)
        else:
            heapq.heappush(self.min_heap, num)
        
        # balance
        if(len(self.min_heap)>len(self.max_heap)):
            top = heapq.heappop(self.min_heap)
            heapq.heappush(self.max_heap, -top)
        
        elif(len(self.max_heap)>len(self.min_heap)+1):
            top = heapq.heappop(self.max_heap)
            heapq.heappush(self.min_heap, -top)

    def findMedian(self) -> float:
        # TC- O(1)
        if len(self.min_heap)==len(self.max_heap):
            return (self.min_heap[0]-self.max_heap[0])/2
        else:
            return -self.max_heap[0]

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()